### PR TITLE
[Intel MKL] Enable 3d shapes in the MklSlice op

### PR DESCRIPTION
**Pull request by wenxizhu**
Support 3d shapes (NCDHW / NDHWC) in MklSliceOp.

Existing MklSliceOp support only 2d shapes and crash all 3d-CNN models such as https://github.com/Youngseok0001/Intel_benchmark/tree/master/3D_CNN.

### CODE REVIEW

File: tensorflow/python/kernel_tests/slice_op_test.py [hunk 1]
```diff
@@ -146,6 +147,19 @@ def testSingleDimension(self):
...
+      input_shape = [1, 8, 8, 8, 3]
+      inp = np.random.rand(*input_shape).astype("f")
+      a = constant_op.constant(inp, shape=input_shape, dtype=dtypes.float32)
+
+      filter_shape = [2, 2, 2, 3, 3]
+      filters = np.random.rand(*filter_shape).astype("f")
+      conv_t = nn_ops.conv3d(a, filter=filters, strides=[1, 1, 1, 1, 1],
+                             padding="SAME")
+      slice_t = array_ops.slice(conv_t, [0, 0, 0, 0, 0], [1, 2, 1, 2, 1])
+      result = self.evaluate(slice_t)
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/slice_op_test.py]: Please add a check if the results are correct.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR!

File: tensorflow/python/kernel_tests/slice_op_test.py (hunk 1 above)
Comment (wenxizhu) [reply tensorflow/python/kernel_tests/slice_op_test.py]: Check code added.

**Review by penpornk — APPROVED** [review]
Thank you!

File: tensorflow/python/kernel_tests/slice_op_test.py [hunk 2]
```diff
@@ -146,6 +147,27 @@ def testSingleDimension(self):
...
+      filters = random_ops.random_normal(filter_shape,
+                                         dtype=dtypes.float32,
+                                         seed=0)
+
+      conv_t = nn_ops.conv3d(inputs,
+                             filter=filters,
+                             strides=[1, 1, 1, 1, 1],
+                             padding="VALID")
+      slice_t = array_ops.slice(conv_t, [0, 1, 1, 1, 0], [1, 1, 1, 1, 8])
+      result = slice_t.eval()
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/slice_op_test.py]: `slice_t.eval()` doesn't work in eager mode. Please use `self.evaluate` instead.
```suggestion
      result = self.evaluate(slice_t.eval)
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Early test results have failures in `Ubuntu Python2` and `Ubuntu Python3`.

**Review by penpornk — DISMISSED** [review]
Thank you!

**Comment by penpornk** [comment]
@wenxizhu Please check the `SliceTest.test3Dimension` failure in the `Ubuntu Python2` [test](https://source.cloud.google.com/results/invocations/ab246701-c3ee-4309-a10e-6be5c3182c01/log).

**Comment by wenxizhu** [comment]
> @wenxizhu Please check the `SliceTest.test3Dimension` failure in the `Ubuntu Python2` [test](https://source.cloud.google.com/results/invocations/ab246701-c3ee-4309-a10e-6be5c3182c01/log).

@penpornk Fixed.

**Review by penpornk — DISMISSED** [review]
Thank you!

**Comment by penpornk** [comment]
@wenxizhu @trollyxia 
Since the last commit is not done by the pull request submitter, could each of you reply with just `@googlebot I consent.` to show your consent?

**Comment by wenxizhu** [comment]
@penpornk I changed the author of the last commit. Is the CLA OK now?

File: tensorflow/python/kernel_tests/slice_op_test.py [hunk 3]
```diff
@@ -28,8 +28,9 @@
 from tensorflow.python.ops import array_ops
 from tensorflow.python.ops import gradients_impl
 from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import nn_ops
 from tensorflow.python.platform import test
-
+from tensorflow.python.ops import random_ops
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/slice_op_test.py]: This line should be above `from tensorflow.python.platform import test`

**Review by penpornk — CHANGES_REQUESTED** [review]
The CLA is okay now. Thank you very much! I have one more minor comment, please.

File: tensorflow/python/kernel_tests/slice_op_test.py (hunk 3 above)
Comment (wenxizhu) [reply tensorflow/python/kernel_tests/slice_op_test.py]: @penpornk OK, fixed.

**Review by penpornk — DISMISSED** [review]
Thank you very much!

**Comment by penpornk** [comment]
@wenxizhu Could you please help take a look at `SliceTest.test3Dimension` in the [Ubuntu CPU test](https://source.cloud.google.com/results/invocations/d7492126-2897-46e6-92a1-d514bba930e5/log)? Thank you!

**Comment by wenxizhu** [comment]
> @wenxizhu Could you please help take a look at `SliceTest.test3Dimension` failure in the [Ubuntu CPU test](https://source.cloud.google.com/results/invocations/d7492126-2897-46e6-92a1-d514bba930e5/log)? Thank you!

@penpornk Could you give me some details of how to reproduce this failure? It passed in our local test environment.

**Comment by penpornk** [comment]
@wenxizhu I can't reproduce it either. I'm thinking it might have to do with how `random_ops.random_normal` is implemented (that somehow make seed=0 get different answer on different OS/architecture). Could you please generate the input as a numpy array first and create the solution with numpy slice instead?

**Comment by wenxizhu** [comment]
> @wenxizhu I can't reproduce it either. I'm thinking it might have to do with how `random_ops.random_normal` is implemented (that somehow make seed=0 get different answer on different OS/architecture). Could you please generate the input as a numpy array first and create the solution with numpy slice instead? This way the solution isn't hard-coded and we don't have to worry about different initialization.

I'm also skeptical about this, which `seed=0` is not a portable across devices (most likely they have different `random()` implementations). I will commit a fix soon.

**Comment by wenxizhu** [comment]
@penpornk Let's see if it works now.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by penpornk — DISMISSED** [review]
I still think using `numpy.random` and Python `slice` should be enough. But we are running out of time, so let's see if this works.

**Comment by penpornk** [comment]
Other tests are still running, but `Ubuntu Sanity` is already finished and it's failing. Please fix the pylint error.

**Review by penpornk — APPROVED** [review]
Thank you, @agramesh1!

File: tensorflow/core/util/mkl_util.h [hunk 4]
```diff
@@ -1117,6 +1117,24 @@ inline memory::dims MklDnnDimsInNCHW(const memory::dims& in_dims,
   return memory::dims({n, c, h, w});
 }
 
+/// Overloaded version of function TFShapeToMklDnnDimsInNCDHW above.
+/// Input parameters are self-explanatory.
+inline memory::dims MklDnnDimsInNCDHW(const memory::dims& in_dims,
+                                      TensorFormat format) {
+  // Validate format.
+  CHECK_NE(TFDataFormatToMklDnnDataFormat(format),
```
Comment (penpornk) [inline tensorflow/core/util/mkl_util.h]: Could you please change this to `DCHECK`?

**Review by penpornk — CHANGES_REQUESTED** [review]

File: tensorflow/core/util/mkl_util.h (hunk 4 above)
Comment (penpornk) [reply tensorflow/core/util/mkl_util.h]: Never mind. I think I'll fix it internally.

Comment (agramesh1) [reply tensorflow/core/util/mkl_util.h]: Thanks

**Comment by gbaned** [comment]
@wenxizhu Could you please resolve the conflicts? Thanks!

**Comment by wenxizhu** [comment]
@penpornk @gbaned I'm a little confused, where these conflicts come from? I didn't change any of there files.

**Comment by penpornk** [comment]
Sorry there is a misunderstanding. This PR has already been merged since yesterday (https://github.com/tensorflow/tensorflow/commit/8c1d5b95c7ed83e6b6c1b9ece06e310777565d27). For some reason it isn't marked so (probably because I made some additional changes internally). Closing the PR.

**Pull request closed without merging**